internal/sdkio/ringbuffer.go

```diff
@@ -0,0 +1,100 @@
+package sdkio
+
+import (
+	"io"
+)
+
+// RingBuffer struct satisfies io.ReadWrite
+// interface.
+//
+// ReadBuffer is a revolving buffer data structure
+// which can be used to store snapshots of data in a
+// revolving window.
+type RingBuffer struct {
+	slice []byte
+	start int
+	end   int
+	size  int
+}
+
+// New method takes in the capacity as an int
+// and returns a RingBuffer.
+func New(capacity int) *RingBuffer {
+	slice := make([]byte, capacity, capacity)
+	ringBuf := RingBuffer{
+		slice: slice,
+		start: 0,
+		end:   0,
+		size:  0,
+	}
+	return &ringBuf
+}
+
+// Write method inserts the elements
+// in the byte slice, and return
+// the number of bytes written and an error
+//
+// This satisfies io.Writer interface.
+func (r *RingBuffer) Write(p []byte) (int, error) {
+
+	for _, b := range p {
+
+		// check if end points to invalid index
+		// we need to circle back
+		if r.end == len(r.slice) {
+			r.end = 0
+		}
+
+		// check if start points to invalid index
+		// we need to circle back
+		if r.start == len(r.slice) {
+			r.start = 0
+		}
+
+		// if ring buffer is filled,
+		// increment the start index
+		if r.size == len(r.slice) {
+			r.size--
+			r.start++
+		}
+
+		r.slice[r.end] = b
+		r.end++
+		r.size++
+	}
+	return r.size, nil
+}
+
+// Read method on RingBuffer returns the read count
+// along with Error encountered while reading.
+//
+// Read copies the data on the ring buffer into
+// the byte slice provided to the method.
+func (r *RingBuffer) Read(p []byte) (int, error) {
```

Comment:
This method probably should decrement the `r.size` count since it is reading bytes out of the buffer. Incrementing  the start cursor